File: python/src/pmsec/tools/uv.py
from __future__ import annotations

import re
from pathlib import Path

from pmsec.util.context import Context
from pmsec.util.extras import apply_extras, read_extras, remove_extras
from pmsec.util.io import write_atomic
from pmsec.util.lines import read_key, remove_key, set_key
from pmsec.util.paths import uv_config_path
from pmsec.util.version import build_preflight

NAME = "uv"
KEY = "exclude-newer"
DOCS = "https://docs.astral.sh/uv/reference/settings/#exclude-newer"
MIN_BIN = (0, 9, 17)
EXTRAS: list[dict] = [
    {"key": "index-strategy", "expected": '"first-index"', "line": 'index-strategy = "first-index"'},
]

preflight = build_preflight(
    NAME, MIN_BIN,
    'writing exclude-newer = "N days" will break this uv until you `uv self update` '
    "(file will fail to parse).",
)

_DURATION = re.compile(r'^"\s*(\d+)\s*(day|days|d|week|weeks|w)\s*"$', re.IGNORECASE)


def path(ctx: Context) -> Path:
    return uv_config_path(ctx.env, ctx.home, ctx.platform)


def _parse_days(value: str | None) -> int | None:
    if value is None:
        return None
    m = _DURATION.match(value)
    if not m:
        return None
    n = int(m.group(1))
    return n * 7 if m.group(2).lower() in ("week", "weeks", "w") else n


def read(ctx: Context) -> dict:
    p = path(ctx)
    raw = p.read_text("utf-8") if p.exists() else ""
    value = read_key(raw, KEY)
    return {"path": str(p), "configured": value, "days": _parse_days(value), "extras": read_extras(raw, EXTRAS)}


def write(days: int, ctx: Context) -> dict:
    p = path(ctx)
    raw = p.read_text("utf-8") if p.exists() else ""
    write_atomic(p, apply_extras(set_key(raw, KEY, f'{KEY} = "{days} days"'), EXTRAS))
    return {"path": str(p)}


def unset(ctx: Context) -> dict:
    p = path(ctx)
    if not p.exists():
        return {"path": str(p), "removed": False}
    before = p.read_text("utf-8")
    after, removed_cooldown = remove_key(before, KEY)
    after, removed_extras = remove_extras(after, EXTRAS)
    removed = removed_cooldown or removed_extras
    if removed:
        write_atomic(p, after)
    return {"path": str(p), "removed": removed}

File: node/src/tools/uv.mjs
import { uvConfigPath } from "../util/paths.mjs";
import { readSafe, writeAtomic } from "../util/io.mjs";
import { readKey, setKey, removeKey } from "../util/lines.mjs";
import { readExtras, applyExtras, removeExtras } from "../util/extras.mjs";
import { buildPreflight } from "../util/version.mjs";

export const name = "uv";
export const key = "exclude-newer";
export const docs = "https://docs.astral.sh/uv/reference/settings/#exclude-newer";
export const minBin = [0, 9, 17];
export const extras = [
  { key: "index-strategy", expected: '"first-index"', line: 'index-strategy = "first-index"' },
];

export function path(ctx) { return uvConfigPath(ctx.env, ctx.home, ctx.platform); }

export const preflight = buildPreflight(name, minBin,
  "writing exclude-newer = \"N days\" will break this uv until you `uv self update` (file will fail to parse).");

function parseDays(value) {
  if (value === null) return null;
  const m = value.match(/^"\s*(\d+)\s*(day|days|d|week|weeks|w)\s*"$/i);
  if (!m) return null;
  const n = Number(m[1]);
  return /^(week|weeks|w)$/i.test(m[2]) ? n * 7 : n;
}

export async function read(ctx) {
  const p = path(ctx);
  const raw = await readSafe(p);
  const value = readKey(raw, key);
  return { path: p, configured: value, days: parseDays(value), extras: readExtras(raw, extras) };
}

export async function write(days, ctx) {
  const p = path(ctx);
  let text = setKey(await readSafe(p), key, `${key} = "${days} days"`);
  text = applyExtras(text, extras);
  await writeAtomic(p, text);
  return { path: p };
}

export async function unset(ctx) {
  const p = path(ctx);
  const before = await readSafe(p);
  const cooldown = removeKey(before, key);
  const ex = removeExtras(cooldown.text, extras);
  const removed = cooldown.removed || ex.removed;
  if (removed) await writeAtomic(p, ex.text);
  return { path: p, removed };
}
